README.md
# cb-taxii-connector (CentOS/RHEL 7/8)

VMware Carbon Black EDR connector for pulling and converting STIX information from TAXII Service Providers into EDR Feeds.

You can install the pre-built RPMs via YUM by using the CB Open Source repository.
*(See CbOpenSource.repo and put that in /etc/yum.repos.d/)*

The pre-built RPM is supported via our [User eXchange (Jive)](https://community.carbonblack.com/community/developer-relations) 
and via email to dev-support@carbonblack.com.  


## Support

1. View all API and integration offerings on the [Developer Network](https://developer.carbonblack.com/) along with reference documentation, video tutorials, and how-to guides.
2. Use the [Developer Community Forum](https://community.carbonblack.com/t5/Developer-Relations/bd-p/developer-relations) to discuss issues and get answers from other API developers in the Carbon Black Community.
3. Report bugs and change requests to [Carbon Black Support](http://carbonblack.com/resources/support/).


## Introduction

This document describes how to install and use the Cb Response TAXII Connector. This connector allows for the importing of STIX data by querying one or more TAXII services and retrieving that data and then converting it into CB feeds using the CB JSON format for IOCs. The job queries for available STIX/TAXII data that is newer than the last time it asked, and by default runs every hour.

For each TAXII service, available “collections” are enumerated and a Cb Response Feed is created. For example, if you have two TAXII services and each exposes two collections, you will have four CB feeds as a result of this connector.

The following IOC types are extracted from STIX data:

* MD5 Hashes
* SHA-256 Hashes
* Domain Names
* IP-Addresses
* IP-Address Ranges


## Requirements

This EDR TAXII Connector has the following requirements:

* *VMware Carbon Black EDR 5.0 (or greater)* – this integration leverages API calls and feed functionality available in Cb Response 5.0 and newer.  In order to check the version, you can run the following rpm command on your server:

```
[root@localhost ~]# rpm -qa | grep cb-enterprise
cb-enterprise-5.0.0.150122.1654-1.el6.x86_64
```

* *Access to TAXII Service Provider* – the purpose of this integration is to retrieve STIX threat information via a TAXII service, so if you do not have access to a TAXII service this integration will be of no value. Example services are SoltraEdge and HailATaxii.com


## Installation

Take the following steps to install the Cb Response Taxii Connector:

1. Install the CbOpenSource.repo file found in the root of this repository (place it in /etc/yum.repos.d/ on your head CB Server node.)
2. Install by issuing the following command as root (or sudo): yum install python-cbtaxii -y


## Upgrades

When an upgrade is available, it should be as easy as doing the following:

    yum install python-cbtaxii -y

* *Note* Upgrading from 1.0/1.1 to 1.2, you'll need to supply auth_token=(CB Server Admin API Token) in the config file under the cbconfig directive.

Please note that a new /etc/cb/integrations/cbtaxii/cbtaxii.conf.example might be made available in some cases, at which point any new settings should be studied, understood, and applied to the production configuration file if necessary.


## Configuration

You’ll need to place a configuration file in the following location:
`/etc/cb/integrations/cbtaxii/cbtaxii.conf`

A sample file is provided in `/etc/cb/integrations/cbtaxii/cbtaxii.conf.example`, so you can rename the file with the following command:
```
mv /etc/cb/integrations/cbtaxii/cbtaxii.conf.example /etc/cb/integrations/cbtaxii/cbtaxii.conf
```

From here, one or more TAXII services can be configured.


## Execution

By default the linux cron daemon will run this integration every day at 1:00 AM to check for new data from the TAXII services you 
have configured. When it runs it will use the current settings found in `/etc/cb/integrations/cbtaxii/cbtaxii.conf`, 
so make sure you are careful when changing any of those settings.

When you first install the connector, you might not want to wait until the hour mark for the job to run. In this case, 
you can force the connector to run manually. As either *root* or *cb* user on the Cb Response Server, execute the 
following command:

```
/usr/share/cb/integrations/cbtaxii/cbtaxii -c /etc/cb/integrations/cbtaxii/cbtaxii.conf
```

It is perfectly fine to do this because the script will only allow one copy of itself to run at a time, so you don’t 
have to worry about the cron daemon attempting to run this while your manual instance is still executing.

*Note #1: this script can take a long time to run depending on the amount of data available from the TAXII services you 
have configured.*

*Note #2: this script logs everything to /var/log/cb/integrations/cbtaxii/cbtaxii.log , so you will see very little 
output when you run it manually.*


You can also enable debug logging by executing:

```
/usr/share/cb/integrations/cbtaxii/cbtaxii -d 
```


## Troubleshooting

If you suspect a problem, please first look at the cbtaxii connector logs found here:

`/var/log/cb/integrations/cbtaxii/cbtaxii.log`

(There might be multiple files as the logger "rolls over" when the log file hits a certain size).


We've seen where Soltra Edge had a user account that wasn't returning data past a particular date for a specific username.  We don't know why this was the case.  The customer created a new SoltraEdge user account and used those credentials in our connector and everything went back to working.

Additionally, due to STIX being a particulary verbose format, sometimes IOCs are stored in fields that we don't expect.  This could result in some IOCs you see in your Taxii platform (such as SoltraEdge) but not show up in Cb Response.  For this and other issues, you can export the raw XML that our connector receives so we can see how information is represented.  To export, use the following command, then contact us and we'll setup a place for you to place the exported XML for our analysis.

```
/usr/share/cb/integrations/cbtaxii/cbtaxii -c /etc/cb/integrations/cbtaxii/cbtaxii.conf --export-dir
```

## Building

The easiest way to build the application is via PyInstaller and utilizing Carbon Black EDRs virtual Python3 enviornment (or a clone of it is actually best).
To do this:

1. Install PyInstaller: ```/usr/share/cb_clone/virtualenv/bin/python –m pip install PyInstaller```
2. Created symbolic links for /bin/python and /bin/pyinstaller via:
```
ln –s "$(which pyinstaller)" /bin/pyinstaller
ln –s /usr/share/cb_clone/virtualenv/bin/python /bin/python
```
3. Build and (re)install:
```
cd </path/to/source/code/directory/here>
rm -rf /home/user/rpm_build/
mkdir /home/user/rpm_build/
/usr/share/cb_clone/virtualenv/bin/python setup.py build_rpm --rpmbuild-dir=/home/user/rpm_build/
yum remove python-cbtaxii.x86_64
yum install /home/user/rpm_build/RPMS/x86_64/python-cbtaxii-1.6.7-4.el8.x86_64.rpm --nogpgcheck
```
4. Run the cbtaxii connector, after configuring it (see 'Configuration' above)
```
/usr/share/cb/integrations/cbtaxii/cbtaxii -c /etc/cb/integrations/cbtaxii/cbtaxii.conf
```
5. Fix errors as necessary (rinse & repeat)
	* If ImportError is reported upon execution, it's likely that one or more packages is missing from either the virtual Python3 enviornment used for building or the binary itself. To solve:

	   	* Verify the package is installed in the virtual Python3 enviornment: 
	   	```/usr/share/cb_clone/virtualenv/bin/python –m pip install <missing package>```
			
	   	* If necessary, add the package to the cb-taxii-connector.spec PyInstaller file: 
	   	```datas.extend([(get_package_paths('<package>')[1], '<package>')])```

	
6. When utilizing a custom build, it may be beneficial to exclude python-cbtaxii from the /etc/yum.conf to prevent accidental upgrades. To do this add the following:
```exclude=<any existing exclusions here> python-cbtaxii*```


If this connector is going to run on Centos 6 then you need to use the source package of lxml from pip.  
Use this command to grab python requirements:
```
pip insall --no-binary lxml -r requirements.txt
```


CHANGELOG.md
CHANGELOG.md
# CB EDR Taxii Connector Changelog

## v1.6.7-4
#### Features
 * Added new config option, ioc_exclusions that supports excluding one or more artifact types from Threat Report creations. Reference the cbtaxii.conf.example for more information on this config option.
 
#### Bug Fixes / Changes
 * Fixed a bug that caused only the first configured TAXII connector connection to be run.
 * Fixed a bug where building via PyInstaller led to a broken cb-taxii-connector binary due to missing packages in the spec file.
 * Now the oldest Threat Reports are truncated, versus the newest, when a Threat Feed reaches its maximum Threat Report size.
 * Now utilizes the time of the fetch for the Threat Report created timestamp as opposed to EPOCH when no timestamp is provided by the indicator itself.
 * Improvements to the cbtaxii.conf.example file for better understanding of config options and defaults.
 * Improvements to the documentation, including documentation of build steps.
 * Slight improvements in error reporting/handling


cbopensource/connectors/taxii/bridge.py
#  coding: utf-8
#  VMware Carbon Black EDR Taxii Connector © 2013-2020 VMware, Inc. All Rights Reserved.
################################################################################

import datetime
import logging
import os
import sys
import tempfile
from typing import Any, AnyStr, Dict, List, Optional, Tuple, Union

import dateutil.tz
from cabby import Client10, Client11, create_client
from cabby.constants import (CB_CAP_11, CB_SMIME, CB_STIX_XML_10, CB_STIX_XML_101, CB_STIX_XML_11, CB_STIX_XML_111,
                             CB_XENC_122002)
from cabby.entities import Collection as CabbyCollection
from cbapi.errors import ServerError
from cbapi.example_helpers import get_object_by_name_or_id
from cbapi.response import CbResponseAPI, Feed
from lxml import etree
from stix.core import STIXPackage

# core packages
from .cb_feed_util import build_feed_data, FeedHelper
from .config_util import parse_config, TaxiiConfigurationException
from .cybox_parse import cybox_parse_observable
from .singleton import SingleInstance, SingleInstanceException
from .util import cleanup_string, dt_to_seconds

CB_STIX_XML_12 = 'urn:stix.mitre.org:xml:1.2'

BINDING_CHOICES = [CB_STIX_XML_111, CB_CAP_11, CB_SMIME, CB_STIX_XML_12,
                   CB_STIX_XML_10, CB_STIX_XML_101, CB_STIX_XML_11,
                   CB_XENC_122002]

EPOCH = datetime.datetime(1970, 1, 1).replace(tzinfo=dateutil.tz.tzutc())

_logger = logging.getLogger(__name__)


class CbTaxiiFeedConverter(object):
    """
    Class to convert TAXII feeds into EDR feeds.
    """

    def __init__(self, config_file_path: str, debug_mode: bool = False, import_dir: str = '',
                 export_dir: Optional[str] = None, strict_mode: bool = False):
        """
        Parse config file and save off the information we need.

        NOTE: At present, import path is unused

        :param config_file_path: configuration file location
        :param debug_mode: If True, operate in debug mode
        :param import_dir: feed import directory
        :param export_dir: export directory (optional)
        :param strict_mode: It True, be harsher wit config
        """
        try:
            config_dict = parse_config(config_file_path, strict_mode=strict_mode)
        except TaxiiConfigurationException as err:
            _logger.error(f"{err}", exc_info=False)
            sys.exit(-1)

        if debug_mode:
            _logger.debug(f"Config: {config_dict}")

        self.server_url = config_dict.get('server_url', 'https://127.0.0.1')
        self.api_token = config_dict.get('api_token', '')
        self.sites = config_dict.get('sites', [])
        self.debug = config_dict.get('debug', False)
        self.export_dir = export_dir
        self.import_dir = import_dir
        self.integration_name = 'Cb Taxii Connector 1.6.5'

        self.http_proxy_url = config_dict.get('http_proxy_url', None)
        self.https_proxy_url = config_dict.get('https_proxy_url', None)

        # if exporting, make sure the directory exists
        if self.export_dir and not os.path.exists(self.export_dir):
            os.mkdir(self.export_dir)

        # Test Cb Response connectivity
        try:
            self.cb = CbResponseAPI(url=self.server_url, token=self.api_token,
                                    ssl_verify=False, integration_name=self.integration_name)
            self.cb.info()
        except Exception as err:
            _logger.error(f"Failed to make connection: {err}", exc_info=True)
            sys.exit(-1)

    @staticmethod
    def write_to_temp_file(message: AnyStr) -> Tuple[tempfile.NamedTemporaryFile, str]:
        """
        Write text to a temp file for later use.

        :param message: text to be saved
        :return: Tuple of (NamedTemporaryFile, tempfile name)
        """
        temp_file = tempfile.NamedTemporaryFile()
        temp_file.write(message)
        temp_file.flush()
        return temp_file, temp_file.name

    # NOTE: currently unused; retained for future need
    # noinspection PyUnusedFunction
    def read_from_xml(self) -> List[str]:
        """
        Walk the import dir and return all filenames.  We are assuming all xml files.

        :return: List of filenames
        """
        the_list = []
        if self.import_dir is None:  # possible if input dir is not specified
            _logger.warning("Input directory was not specified -- skipping xml read")
            return the_list

        for (dirpath, dirnames, filenames) in os.walk(self.import_dir):
            the_list.extend(filenames)
            break
        return the_list

    def export_xml(self, feed_name: str, start_time: str, end_time: str, block_num: int, message: AnyStr) -> None:
        """
        :param feed_name: name of the feed, for the holding directory name
        :param start_time: start time
        :param end_time: end time
        :param block_num: write block number (for uniqueness)
        :param message: feed text
        """
        # create a directory to store all content blocks
        dir_name = f"{feed_name}".replace(' ', '_')
        full_dir_name = os.path.join(self.export_dir, dir_name)

        # Make sure the directory exists
        if not os.path.exists(os.path.join(self.export_dir, dir_name)):
            os.mkdir(full_dir_name)

        # Actually write the file
        file_name = f"{start_time}-{end_time}-{block_num}".replace(' ', "_")
        full_file_name = os.path.join(full_dir_name, file_name)

        with open(full_file_name, 'wb') as file_handle:
            file_handle.write(message)

    def _import_collection(self, client: Union[Client10, Client11], site: dict, collection: CabbyCollection,
                           data_set: bool = False) -> int:
        """
        Import a taxii client collectio into a feed.

        :param client: Taxii spec client v1.0 or v1.1
        :param site: site definition
        :param collection: cabby collection
        :param data_set: True if DATA_SET, False otherwise
        :return: the EDR feed id, or -1 if not available
        """
        global BINDING_CHOICES

        collection_name = collection.name
        display_name = f"{site.get('site')} {collection_name}"
        sanitized_feed_name = cleanup_string(site.get('site') + collection_name)
        feed_summary = display_name
        available = collection.available
        collection_type = collection.type
        default_score = site.get('default_score')
        _logger.info(f"Working on SITE {site.get('site')}, NAME {collection_name}, FEED {sanitized_feed_name}, "
                     f"AVAIL {available}, TYPE {collection_type}")
        _logger.info('-' * 80)

        # if not available, nothing else to do
        if not available:
            return -1

        # Sanity check on start date; provide a bare minimum
        start_date_str = site.get('start_date')
        if not start_date_str or len(start_date_str) == 0:
            start_date_str = "2019-01-01 00:00:00"

        # Create a feed helper object
        feed_helper = FeedHelper(site.get('output_path'), sanitized_feed_name, site.get('minutes_to_advance'),
                                 start_date_str, reset_start_date=site.get('reset_start_date', False))

        if not data_set:
            _logger.info("Feed start time %s" % feed_helper.start_date)

        #
        # Build up the URI for polling
        #

        if not site.get('poll_path', ''):
            uri: Optional[str] = None
        else:
            uri: str = ''
            if site.get('use_https'):
                uri += 'https://'
            else:
                uri += 'http://'

            uri += site.get('site')
            uri += site.get('poll_path')
            _logger.info(f'Poll path: {uri}')

        # build up all the reports for the feed
        reports: List[Dict[str, Any]] = []
        while True:
            num_times_empty_content_blocks = 0
            try:
                try:
                    _logger.info(f"Polling Collection: {collection.name} ...")
                    content_blocks = client.poll(collection_name=collection.name, begin_date=feed_helper.start_date,
                                                 end_date=feed_helper.end_date, content_bindings=BINDING_CHOICES,
                                                 uri=uri)
                except Exception as e:
                    _logger.info(f"{e}")
                    content_blocks = []

                #
                # Iterate through all content_blocks
                #
                num_blocks = 0

                if not data_set:
                    _logger.info(f" ... start_date: {feed_helper.start_date}, end_date: {feed_helper.end_date}")
                for block in content_blocks:
                    _logger.debug(block.content)

                    #
                    # if in export mode then save off this content block
                    #
                    if self.export_dir:
                        self.export_xml(collection_name,
                                        feed_helper.start_date,
                                        feed_helper.end_date,
                                        num_blocks,
                                        block.content)

                    #
                    # This code accounts for a case found with ThreatCentral.io where the content is url encoded.
                    # etree.fromstring can parse this data.
                    #
                    try:
                        root = etree.fromstring(block.content)
                        content = root.find('.//{http://taxii.mitre.org/messages/taxii_xml_binding-1.1}Content')
                        if content is not None and len(content) == 0 and len(list(content)) == 0:
                            #
                            # Content has no children.  So lets make sure we parse the xml text for content and re-add
                            # it as valid XML so we can parse
                            #
                            new_stix_package = etree.fromstring(root.find(
                                "{http://taxii.mitre.org/messages/taxii_xml_binding-1.1}"
                                "Content_Block/{http://taxii.mitre.org/messages/taxii_xml_binding-1.1}Content").text)
                            content.append(new_stix_package)

                        #
                        # Since we modified the xml, we need create a new xml message string to parse
                        #
                        message = etree.tostring(root)

                        #
                        # Write the content block to disk so we can parse with python stix
                        #
                        file_handle, file_path = self.write_to_temp_file(message)

                        #
                        # Parse STIX data
                        #
                        stix_package = STIXPackage.from_xml(file_path)

                        #
                        # if it is a DATA_SET make feed_summary from the stix_header description
                        # NOTE: this is for RecordedFuture, also note that we only do this for data_sets.
                        #       to date I have only seen RecordedFuture use data_sets
                        #
                        if data_set and stix_package.stix_header and stix_package.stix_header.descriptions:
                            for desc in stix_package.stix_header.descriptions:
                                feed_summary = f"{desc.value}: {collection_name}"
                                break

                        #
                        # Get the timestamp of the STIX Package so we can use this in our feed
                        #
                        timestamp = dt_to_seconds(stix_package.timestamp)

                        # check for empty content in this block; we break out after 10 empty blocks
                        if not stix_package.indicators and not stix_package.observables:
                            num_times_empty_content_blocks += 1
                            if num_times_empty_content_blocks > 10:
                                break

                        # Go through all STIX indicators
                        if stix_package.indicators:
                            for indicator in stix_package.indicators:

                                if not indicator or not indicator.observable:
                                    continue

                                if indicator.confidence:
                                    if str(indicator.confidence.value).isdigit():
                                        #
                                        # Get the confidence score and use it for our score
                                        #
                                        score = int(indicator.confidence.to_dict().get("value", default_score))
                                    else:
                                        if str(indicator.confidence.value).lower() == "high":
                                            score = 75
                                        elif str(indicator.confidence.value).lower() == "medium":
                                            score = 50
                                        elif str(indicator.confidence.value).lower() == "low":
                                            score = 25
                                        else:
                                            score = default_score
                                else:
                                    score = default_score

                                if not indicator.timestamp:
                                    timestamp = 0
                                else:
                                    timestamp = int((indicator.timestamp - EPOCH).total_seconds())

                                # Cybox observable returns a list
                                reports.extend(cybox_parse_observable(indicator.observable, indicator, timestamp,
                                                                      score))

                        #
                        # Now lets find some data.  Iterate through all observables and parse
                        #
                        if stix_package.observables:
                            for observable in stix_package.observables:
                                if not observable:
                                    continue

                                # Cybox observable returns a list
                                reports.extend(cybox_parse_observable(observable, None, timestamp, default_score))

                        #
                        # Delete our temporary file
                        #
                        file_handle.close()

                        # increase block count
                        num_blocks += 1
                    except Exception as e:
                        _logger.info(f"{e}")
                        continue

                _logger.info(f"content blocks read: {num_blocks}")
                _logger.info(f"current number of reports: {len(reports)}")

                if len(reports) > site.get('reports_limit'):
                    _logger.info(f"We have reached the reports limit of {site.get('reports_limit')}")
                    break
            except Exception as e:
                _logger.info(f"{e}")

            # If it is just a data_set, the data is unordered, so we can just break out of the while loop
            if data_set:
                break

            if feed_helper.advance():
                continue
            else:
                break

        _logger.info(f"Found {len(reports)} new reports.")

        if not data_set:
            # We only want to concatenate if we are NOT a data set, otherwise we want to refresh all the reports
            _logger.info("Adding existing reports...")
            reports = feed_helper.load_existing_feed_data() + reports

        _logger.info(f"Total number of reports: {len(reports)}")

        if site.get('reports_limit') < len(reports):
            _logger.info("Truncating reports to length {0}".format(site.get('reports_limit')))
            reports = reports[:site.get('reports_limit')]

        try:
            use_icon = site.get('icon_link')
            if not os.path.exists(use_icon):
                _logger.warning(f"Unable to find feed icon at path {use_icon}")
                use_icon = None
            display = f"{site.get('site')} {collection_name}"
            data = build_feed_data(sanitized_feed_name,
                                   display_name,
                                   feed_summary,
                                   site.get('site'),
                                   use_icon,
                                   reports)
        except Exception as err:
            _logger.warning(f"Failed to create feed data for {sanitized_feed_name}: {err}")
            return -1

        if feed_helper.write_feed(data):
            feed_helper.save_details()

        #
        # Create Cb Response Feed if necessary
        #

        feed_id = None

        try:
            feeds = get_object_by_name_or_id(self.cb, Feed, name=sanitized_feed_name)

            if not feeds:
                _logger.info(f"Feed {sanitized_feed_name} was not found, so we are going to create it...")

            elif len(feeds) > 1:
                _logger.warning(f"Multiple feeds found, selecting Feed id {feeds[0].id}")
                feed_id = feeds[0].id

            elif feeds:
                feed_id = feeds[0].id
                _logger.info(f"Feed `{sanitized_feed_name}` was found as Feed ID {feed_id}")

        except Exception as e:
            _logger.info(f"{e}")

        if not feed_id:
            _logger.info(f" ... creating {sanitized_feed_name} feed for the first time")

            f = self.cb.create(Feed)
            f.feed_url = "file://" + feed_helper.path
            f.enabled = site.get('feeds_enable')
            f.use_proxy = False
            f.validate_server_cert = False
            try:
                f.save()
            except ServerError as se:
                if se.error_code == 500:
                    _logger.warning(" ... could not add feed:")
                    _logger.warning("   Received error code 500 from server. This is usually because "
                                    "the server cannot retrieve the feed.")
                    _logger.warning("   Check to ensure the Cb server has network connectivity "
                                    "and the credentials are correct.")
                    _logger.warning("!" * 80 + "\n")
                else:
                    info = feed_helper.dump_feedinfo()
                    _logger.info(f" ... Could not add feed: {se}\n >> {info}")
            except Exception as e:
                info = feed_helper.dump_feedinfo()
                _logger.warning(f" ... Could not add feed: {e}\n >> {info}")
                _logger.warning("!" * 80 + "\n")
            else:
                _logger.info(f"Feed data: {f}")
                _logger.info(f"Added feed. New feed ID is {f.id}")
                feed_id = f.id

        return feed_id

    def perform(self) -> None:
        """
        Perform the taxii hailing service.
        """
        for site in self.sites:
            client: Union[Client10, Client11] = create_client(site.get('site'),
                                                              use_https=site.get('use_https'),
                                                              discovery_path=site.get('discovery_path'))

            #
            # Set verify_ssl and ca_cert inside the client
            #
            client.set_auth(verify_ssl=site.get('ssl_verify'), ca_cert=site.get('ca_cert'))

            #
            # Proxy Settings
            #
            proxy_dict = dict()

            if self.http_proxy_url:
                _logger.info(f"Found HTTP Proxy: {self.http_proxy_url}")
                proxy_dict['http'] = self.http_proxy_url

            if self.https_proxy_url:
                _logger.info(f"Found HTTPS Proxy: {self.https_proxy_url}")
                proxy_dict['https'] = self.https_proxy_url

            if proxy_dict:
                client.set_proxies(proxy_dict)

            # If a username is supplied use basic authentication
            if site.get('username') or site.get('cert_file'):
                _logger.info("Found Username in config, using basic auth...")
                client.set_auth(username=site.get('username'),
                                password=site.get('password'),
                                verify_ssl=site.get('ssl_verify'),
                                ca_cert=site.get('ca_cert'),
                                cert_file=site.get('cert_file'),
                                key_file=site.get('key_file'))

            if not site.get('collection_management_path', ''):
                collections = client.get_collections()
            else:
                uri = ''
                if site.get('use_https'):
                    uri += 'https://'
                else:
                    uri += 'http://'

                uri += site.get('site')
                uri += site.get('collection_management_path')
                _logger.info('Collection Management Path: {}'.format(uri))

                collections: List[CabbyCollection] = client.get_collections(uri=uri)

            if len(collections) == 0:
                _logger.info('Unable to find any collections.  Exiting...')
                sys.exit(0)

            _logger.info("=" * 80)
            for collection in collections:
                _logger.info(f'Collection Name: {collection.name}, Collection Type: {collection.type}')
            _logger.info("=" * 80 + "\n")

            desired_collections = [x.strip() for x in site.get('collections').lower().split(',')]

            want_all = False
            if '*' in desired_collections:
                want_all = True

            for collection in collections:
                if collection.type != 'DATA_FEED' and collection.type != 'DATA_SET':
                    continue

                if collection.type == 'DATA_SET':
                    data_set = True
                else:
                    data_set = False

                if want_all or collection.name.lower() in desired_collections:
                    self._import_collection(client, site, collection, data_set)


################################################################################
# Local utilities
################################################################################


def runner(configpath: str, debug_mode: bool, import_dir: str, export_dir: Optional[str],
           strict_mode: bool = False) -> bool:
    """
    Function to perform the single taxii service.

    :param configpath: path to the taxii config file
    :param debug_mode: If True, execute in debug mode
    :param import_dir: path to the import dir
    :param export_dir: path to the export dir (optional)
    :param strict_mode: If True, be harsher with configuration checking
    :return: True if executed without problems
    """
    try:
        # Setting nice inside script so we don't get killed by OOM
        os.nice(1)

        #
        # run only one instance of this script
        #
        # noinspection PyUnusedLocal
        me = SingleInstance()
        cbt = CbTaxiiFeedConverter(configpath, debug_mode, import_dir, export_dir, strict_mode)
        cbt.perform()
    except SingleInstanceException as e:
        _logger.error(f"Cannot run multiple copies of this script: {e}")
        return False
    except Exception as e:
        _logger.error(f"{e}")
        return False
    return True
